### PR TITLE
meta: support top-level `provenance` keyword for on-prem support

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----

The `provenance` key in snapcraft.yaml can now be moved from underneath `passthrough` to a top-level key.

(CRAFT-1423)

### CODE REVIEW

File: tests/unit/test_projects.py [hunk 1]
```diff
@@ -1148,31 +1187,6 @@ def test_grammar_syntax_error(self, project_yaml_data):
         with pytest.raises(errors.ProjectValidationError, match=error):
             GrammarAwareProject.validate_grammar(data)
 
-    @pytest.mark.parametrize(
```
Comment (mr-cal) [inline tests/unit/test_projects.py:1151]: Unrelated rearrangement - these 2 tests were in the wrong place.

**Comment by mr-cal** [comment]
Spread failure unrelated (https://github.com/snapcore/snapcraft/pull/3962)

File: snapcraft/projects.py [hunk 2]
```diff
@@ -547,6 +548,16 @@ def _validate_architecture_data(cls, architectures):
         """Validate architecture data."""
         return _validate_architectures(architectures)
 
+    @pydantic.validator("provenance")
+    @classmethod
+    def _validate_provenance(cls, provenance):
+        if provenance and not re.match(r"^[a-zA-Z0-9-]+$", provenance):
+            raise ValueError(
+                "provenance must consist of alphanumeric characters and hyphens."
```
Comment (facundobatista) [inline snapcraft/projects.py:556]: maybe "alphanumeric characters and/or hyphens"?

**Review by facundobatista — APPROVED** [review]
Awesome! +1, but check comment, thanks!!

File: snapcraft/projects.py (hunk 2 above)
Comment (mr-cal) [reply snapcraft/projects.py]: Good idea, that's more accurate.

**Comment by mr-cal** [comment]
Merging with 1 unrelated spread failure (resolved by https://github.com/snapcore/snapcraft/pull/3962)

**Review by cmatsuoka — APPROVED** [review]
Looks good.

**Pull request merged**